decrypt issue during spring.cloud.config.server.bootstrap

**Describe the bug**

migrate from
springboot 2.7.x
springcloud 2021.0.6

to 

springboot 3.0.6
springcloud 2022.0.2

set `spring.cloud.config.server.bootstrap` to `true`, configserver no longer decrypt the entry inside the `search-locations` when configure itself. It working fine when change it to plain text.

P.s. it working fine during springboot 2.7.x

bootstrap.yml
```
spring:
  application:
    name: configserver
  profiles:
    active: dev
    default: dev
    include: composite
  cloud:
    config:
      server: ## ConfigServerAutoConfiguration ##
        bootstrap: true
        composite:
          - type: native
            search-locations: 
            - file:/app/config/centralconfig/conf

encrypt:
  key-store:
    location: classpath:/encrypt/encrypt.pkcs12
    password: xxxxxx
    alias: encrypt-dev
    type: pkcs12
```

configserver-dev.yml
```
spring:
  security:
    user:
      name: admin
      # admin
      password: '{cipher}{key:encrypt-dev}AQCMdfP7wlVsVq43iJxdqnPiFyBAzpdPyXd7emIppONCqs6ilazJ8ttZLkmR7x26hAhAIxizFSy3+uYwNjakrm3o4M1MGmKEYJcMbdJKXNOKftIf15qcB0CwokXl2pSPeYq23aUwryBld6wKuBWjVg8mQ+8InYxJUVMTVbrOxg+yK/PCjCcf/j8g3g7PtTARaRo2Wz9lnojRZNFs7aHXDdyzUH6USrxFZ7l4kb6eficyhMyv+he7l8LOCckSQIBdEHsoeUnLCfAfSAm2xbbU47H8fSK5058MxJWRuSDSMaoh2O/+UcKUs38WPbofTQGudaxsn0i8xJAvMuTViEZBy+sxIl0raIYdzTcb9Ywy3DTmK5Hj6YGj4k1qiv14MQbHNsk='
      roles: USER, ADMIN
```

the key is to reproduce is:

- set `spring.cloud.config.server.bootstrap` to `true`
- using `composite` profile with `native` type
- since configserver using `configserver ` as `spring.application.name`, then inside `search-location` was contain `configserver.yml`. This file was contain some congiuration which required to startup configserver itself, like the spring security user password as shown above at my 1st post, server.ssl as show below  .E.g.

   ```
    server:
      ssl:
        key-store: file:/app/tls/micro/application-key.pkcs12
        key-store-password: '{cipher}{key:encrypt-dev}AQCJAj1KyOVG7Pz4E6+0bvKi02gM37f5A7/FU+7J6ZJfAbP1moYUEVG8pc4tFep3dzw+GJKsVqDgWFfq56eCcU6TQ4rYxf6sietO0goWYSC3bsymoH46DuGjP1LunzHGSZMGzURInTfGfU2ag027MRiiqiZkaNPg8uU18Aqt0cdf1ziHDzgj3WlD6IyQ0fknpUMKH1dMR1380NPvfYrZ25t8BkFOzfAc8vkNnn96T3I428MynRBkZZQU8IDxWrJ+JpXFE/20P41yTwLkrZryRgjHe+iBS6C9XpLCzlufWkDUKusny+Vgghwq4uOYwjnDA8xwcicJo+EbQdggZn/Sgsw3dKUdHQ83Q8OEy1pRt1hOL/vZuVBfBQOM9WZIOB7tQ50='
        key-store-type: pkcs12
        key-password: '{cipher}{key:encrypt-dev}AQCJAj1KyOVG7Pz4E6+0bvKi02gM37f5A7/FU+7J6ZJfAbP1moYUEVG8pc4tFep3dzw+GJKsVqDgWFfq56eCcU6TQ4rYxf6sietO0goWYSC3bsymoH46DuGjP1LunzHGSZMGzURInTfGfU2ag027MRiiqiZkaNPg8uU18Aqt0cdf1ziHDzgj3WlD6IyQ0fknpUMKH1dMR1380NPvfYrZ25t8BkFOzfAc8vkNnn96T3I428MynRBkZZQU8IDxWrJ+JpXFE/20P41yTwLkrZryRgjHe+iBS6C9XpLCzlufWkDUKusny+Vgghwq4uOYwjnDA8xwcicJo+EbQdggZn/Sgsw3dKUdHQ83Q8OEy1pRt1hOL/vZuVBfBQOM9WZIOB7tQ50='
        trust-store: file:/app/tls/micro/application-trust.pkcs12
        trust-store-password: '{cipher}{key:encrypt-dev}AQCJAj1KyOVG7Pz4E6+0bvKi02gM37f5A7/FU+7J6ZJfAbP1moYUEVG8pc4tFep3dzw+GJKsVqDgWFfq56eCcU6TQ4rYxf6sietO0goWYSC3bsymoH46DuGjP1LunzHGSZMGzURInTfGfU2ag027MRiiqiZkaNPg8uU18Aqt0cdf1ziHDzgj3WlD6IyQ0fknpUMKH1dMR1380NPvfYrZ25t8BkFOzfAc8vkNnn96T3I428MynRBkZZQU8IDxWrJ+JpXFE/20P41yTwLkrZryRgjHe+iBS6C9XpLCzlufWkDUKusny+Vgghwq4uOYwjnDA8xwcicJo+EbQdggZn/Sgsw3dKUdHQ83Q8OEy1pRt1hOL/vZuVBfBQOM9WZIOB7tQ50='
        trust-store-type: pkcs12
   ```

Then it will endup not able to start due to `keystore password was incorrect`, because configserver didn't decrypt the cipher when bootrap itself. 
It working fine if i change the all cipher (only the configserver related required to startup itself) to plain text. After it able to start with plain text, any `http://localhost:8888/application/default` or `http://localhost:8888/<application>/<profile>` it working fine because it was runtime decryption

```
org.springframework.context.ApplicationContextException: Unable to start web server
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:164) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:602) ~[spring-context-6.0.11.jar:6.0.11]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:732) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:434) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:310) ~[spring-boot-3.0.10.jar:3.0.10]
        at io.cv.configserver.ConfigServerApplication.main(ConfigServerApplication.java:24) ~[main/:na]
Caused by: java.lang.IllegalStateException: org.springframework.boot.web.server.WebServerException: Could not load key store 'file:/app/tls/micro/application-key.pkcs12'
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.getKeyManagers(SslBuilderCustomizer.java:127) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.customize(SslBuilderCustomizer.java:77) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.UndertowWebServerFactoryDelegate.createBuilder(UndertowWebServerFactoryDelegate.java:167) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.UndertowServletWebServerFactory.getWebServer(UndertowServletWebServerFactory.java:297) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:183) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:161) ~[spring-boot-3.0.10.jar:3.0.10]
        ... 6 common frames omitted
Caused by: org.springframework.boot.web.server.WebServerException: Could not load key store 'file:/app/tls/micro/application-key.pkcs12'
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.loadStore(SslBuilderCustomizer.java:199) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.loadKeyStore(SslBuilderCustomizer.java:171) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.getKeyStore(SslBuilderCustomizer.java:145) ~[spring-boot-3.0.10.jar:3.0.10]
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.getKeyManagers(SslBuilderCustomizer.java:112) ~[spring-boot-3.0.10.jar:3.0.10]
        ... 11 common frames omitted
Caused by: java.io.IOException: keystore password was incorrect
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2159) ~[na:na]
        at java.base/sun.security.util.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:221) ~[na:na]
        at java.base/java.security.KeyStore.load(KeyStore.java:1473) ~[na:na]
        at org.springframework.boot.web.embedded.undertow.SslBuilderCustomizer.loadStore(SslBuilderCustomizer.java:195) ~[spring-boot-3.0.10.jar:3.0.10]
        ... 14 common frames omitted
Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
        ... 18 common frames omitted
```

--- a/spring-cloud-config-server/src/main/resources/META-INF/spring.factories
+++ b/spring-cloud-config-server/src/main/resources/META-INF/spring.factories
@@ -1,8 +1,8 @@
 # Bootstrap components
 org.springframework.cloud.bootstrap.BootstrapConfiguration=\
 org.springframework.cloud.config.server.bootstrap.ConfigServerBootstrapConfiguration,\
-org.springframework.cloud.config.server.config.DefaultTextEncryptionAutoConfiguration,\
 org.springframework.cloud.config.server.config.RsaEncryptionAutoConfiguration,\
+org.springframework.cloud.config.server.config.DefaultTextEncryptionAutoConfiguration,\
 org.springframework.cloud.config.server.config.EncryptionAutoConfiguration
 
 # Environment PostProcessor

--- a/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/encryption/EncryptionAutoConfigurationTests.java
+++ b/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/encryption/EncryptionAutoConfigurationTests.java
@@ -1,0 +1,127 @@
+/*
+ * Copyright 2002-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.config.server.encryption;
+
+import org.junit.jupiter.api.Test;
+
+import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration;
+import org.springframework.boot.autoconfigure.web.ServerProperties;
+import org.springframework.boot.autoconfigure.web.servlet.ServletWebServerFactoryAutoConfiguration;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.cloud.bootstrap.encrypt.TextEncryptorUtils;
+import org.springframework.cloud.config.server.config.DefaultTextEncryptionAutoConfiguration;
+import org.springframework.cloud.config.server.config.RsaEncryptionAutoConfiguration;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.security.crypto.encrypt.Encryptors;
+import org.springframework.security.crypto.encrypt.TextEncryptor;
+import org.springframework.security.rsa.crypto.RsaSecretEncryptor;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Ryan Baxter
+ */
+public class EncryptionAutoConfigurationTests {
+
+	@Test
+	public void defaultNoKeyAutoConfigurationTest() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class).properties("server.port=0").run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		String[] textEncryptorLocatorNames = context.getBeanNamesForType(TextEncryptorLocator.class);
+		assertThat(textEncryptor).isInstanceOf(TextEncryptorUtils.FailsafeTextEncryptor.class);
+		assertThat(textEncryptorLocatorNames).isEmpty();
+		context.close();
+	}
+
+	@Test
+	public void defaultKeyStoreAutoConfigurationTest() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class)
+						.properties("server.port=0", "encrypt.key-store.location=classpath:server.jks",
+								"encrypt.key-store.password=letmein", "encrypt.key-store.alias=myKey")
+						.run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		TextEncryptorLocator textEncryptorLocator = context.getBean(TextEncryptorLocator.class);
+		assertThat(textEncryptor).isInstanceOf(RsaSecretEncryptor.class);
+		assertThat(textEncryptorLocator).isInstanceOf(KeyStoreTextEncryptorLocator.class);
+		context.close();
+	}
+
+	@Test
+	public void defaultKeyAutoConfigurationTest() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class).properties("server.port=0", "encrypt.key=mykey").run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		String[] textEncryptorLocatorNames = context.getBeanNamesForType(TextEncryptorLocator.class);
+		assertThat(textEncryptor.getClass().getName())
+				.isEqualTo("org.springframework.security.crypto.encrypt.HexEncodingTextEncryptor");
+		assertThat(textEncryptorLocatorNames).isEmpty();
+		context.close();
+	}
+
+	@Test
+	public void bootstrapNoKeyAutoConfigurationTest() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class)
+						.properties("server.port=0", "spring.cloud.bootstrap.enabled=true").run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		String[] textEncryptorLocatorNames = context.getBeanNamesForType(TextEncryptorLocator.class);
+		assertThat(textEncryptor.getClass().isInstance(Encryptors.noOpText().getClass()));
+		assertThat(textEncryptorLocatorNames).isEmpty();
+		context.close();
+	}
+
+	@Test
+	public void bootstrapKeyAutoConfigurationTests() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class)
+						.properties("server.port=0", "spring.cloud.bootstrap.enabled=true", "encrypt.key=mykey").run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		assertThat(textEncryptor.getClass().getName())
+				.isEqualTo("org.springframework.security.crypto.encrypt.HexEncodingTextEncryptor");
+		context.close();
+	}
+
+	@Test
+	public void bootstrapKeyStoreAutoConfigurationTest() {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(
+				DefaultTextEncryptionAutoConfiguration.class, RsaEncryptionAutoConfiguration.class,
+				ServletWebServerFactoryAutoConfiguration.class, ServerProperties.class,
+				PropertyPlaceholderAutoConfiguration.class)
+						.properties("server.port=0", "spring.cloud.bootstrap.enabled=true",
+								"encrypt.key-store.location=classpath:server.jks", "encrypt.key-store.password=letmein",
+								"encrypt.key-store.alias=myKey")
+						.run();
+		TextEncryptor textEncryptor = context.getBean(TextEncryptor.class);
+		TextEncryptorLocator textEncryptorLocator = context.getBean(TextEncryptorLocator.class);
+		assertThat(textEncryptor).isInstanceOf(LocatorTextEncryptor.class);
+		assertThat(textEncryptorLocator).isInstanceOf(KeyStoreTextEncryptorLocator.class);
+		context.close();
+	}
+
+}

